### ISSUE TITLE
Implement `cp.Expression.__rpow__`

### ISSUE THREAD

**Issue by PTNobel** [cvxpy/cvxpy#1700]
This would allow constraints or objectives of the form `a**x`.

In #1648, we discovered that `__rpow__` is missing. @rileyjmurray proposed a quick sketch of how to implement it:

```python
def expr_rpow(self, other):
    if hasattr(other, 'is_constant'):
        assert other.is_constant()
    if hasattr(other, 'value'):
        other = other.value
    if not isinstance(other, np.ndarray):
        other = np.asarray(other)
    assert np.all(other > 0)
    log_other = np.log(other)
    return cp.exp(cp.multiply(log_other, self))
```

I should have time to open a pull request in the next two or three weeks with some tests and this implementation, but wanted to open an issue first to see if there's any other cases to consider or objections.

I'm thinking we should also add as an atom ```cp.power(const, variable)``` with curvature convex, sign positive, and monotonicity based on the magnitude of the const and a total domain.

Anyways, I'm happy do this towards the end of Stanford's spring break and/or the week following, but figured I'd gather consensus here first.

**Comment by SteveDiamond** [comment]
Definitely worth implementing this, but you need to consider the case where ``other`` is a parameter.

**Comment by PTNobel** [comment]
Question: do parameters have a > 1 constraint I could consider in picking the monotoncity?

**Comment by SteveDiamond** [comment]
Sadly no, parameters have the same attributes as variables, listed here: https://www.cvxpy.org/tutorial/advanced/index.html#attributes

On the other hand, we have discussed adding a ``range`` attribute to restrict things to an interval.

**Comment by rileyjmurray** [comment]
@PTNobel  This expression can't be DPP when ``other`` is a Parameter. The rewriting with "exp" requires taking a log, and DPP doesn't allow nonlinear functions of Parameters:
```
>>> p = cp.Parameter(shape=(2,))
>>> x = cp.Variable(shape=(2,))
>>> y = cp.power(p, 2) @ x
>>> y.is_dpp()
False
>>> 
```
The implementation would need to take the same approach as ``kron``, which suitably coerces a Parameter into a Constant  at canonicalization time.

**Comment by PTNobel** [comment]
I'm not familiar enough with DPP rules to answer this, but would it be possible if we wrote `p**x` in terms of something other than `exp(x * log(p))` to properly parameterize?

I think this is a question of whether a cone program can be affinely related to the parameters values when only using SCS's cones?

**Comment by rileyjmurray** [comment]
I'm confident that there is no formulation that's linear in ``p``, and so no way to make DPP work for this situation. @PTNobel take a look at ``cvxpy/atoms/affine/kron.py`` for how it handles parameters when checking DCP/DPP rules.

**Comment by PTNobel** [comment]
Awesome. Will do. Thanks Riley!

**Comment by PTNobel** [comment]
Barring anything else anyone suggests, I'll start working on this tonight!

**Comment by PTNobel** [comment]
Is there an elegant way to overload cp.power to choose between `cp.power(b: cp.Constant | cp.Parameter, x: cp.Variable)` and `cp.power(x: cp.Variable, p: cp.Constant | cp.Parameter)` there's very little that needs to be shared between the implementation, unlike the various version of $x^p$, and I'm wondering if it would be reasonable to have something like:

```python
def power(b_or_x, x_or_p):
     # Logic to tell which of b and p are const
     if b_is_const:  # Prefer the right variable power atom for
                           # all constant powers because it doesn't
                           # require rational approximation of consts
             return RightVariablePowerAtom(b_or_x, x_or_p)
     elif p_is_const:
             return OldPowerAtom(b_or_x, x_or_p)
     else:
             raise DCPError("f(x)^g(x) is not generally convex")
```

and then two classes to implement the atoms.

(types aren't strict, but hopefully get my point across).

**Comment by PTNobel** [comment]
Obviously this proposal breaks all code with an 

```python
assert type(cp.power) is type
```
and that might be considered a violation of SemVer? I'm not sure what our public API is.

If we want to keep that assert passing, I think the most elegant thing would be to have a:
```PowerAtom``` all of whom's methods just defers to the methods of `PowerAtom.inner` which is one of `RightVariablePowerAtom` or `OldPowerAtom` from my previous message.

**Comment by SteveDiamond** [comment]
We do tend to handle cases like this as you've suggested, with a function that dispatches to two different atoms (see multiplication and quad_form). I think you could maintain the type assertion very easily though even with a dispatch function by making ``LeftPower`` and ``RightPower`` subclasses of an abstract ``power`` base class.

**Comment by PTNobel** [comment]
I guess my question is: am I required to maintain the type assertion by our SemVer rules?

Python+SemVer always confuses me since I'm never sure what's part of the API and what's just an implementation detail.

**Comment by rileyjmurray** [comment]
@PTNobel you're right to point out that it's not clear what is and isn't part of the public API. We should work towards doing a better job of that. In this case, Steven's suggestion of having ``Power`` remain a class in its own right and then dispatch to new subclasses ``LeftPower`` and ``RightPower`` gets the job done within reason. It's true that ``assert type(cp.power) is type`` would fail, but ``assert isinstance(cp.power(val1, val2), type)`` would succeed.
